### ISSUE TITLE
0.8.2 Nav appears to break url pathname callbacks

### ISSUE THREAD

**Issue by daniel-saunders** [facultyai/dash-bootstrap-components#310]
- dash version: `#1.8.0`
- dash-bootstrap-components version: `#0.8.2`
- components affected by bug: Nav

Appears as if Navlink clicks aren't triggering URL path change callback.
The URL appears to update in the display bar (testing in Firefox), but the callback for URL path change doesn't get triggered (although can be triggered if forcing the page to refresh). 

Downgrading to dbc 0.8.1 solved the issue. 

Posting here just for awareness.

**Comment by tcbegley** [comment]
Hi @daniel-saunders 

Thanks for reporting! I wasn't able to reproduce this. I tried the below code on both Chrome and Firefox with dbc 0.8.2 and didn't have any problems.

```python
import dash
import dash_bootstrap_components as dbc
import dash_core_components as dcc
import dash_html_components as html
from dash.dependencies import Input, Output

app = dash.Dash(external_stylesheets=[dbc.themes.BOOTSTRAP])

app.layout = dbc.Container(
    [
        dcc.Location(id="url"),
        dbc.Nav(
            [
                dbc.NavLink("Page 1", href="/page-1"),
                dbc.NavLink("Page 2", href="/page-2"),
            ]
        ),
        html.Div(id="output"),
    ]
)


@app.callback(Output("output", "children"), [Input("url", "pathname")])
def on_url_change(pathname):
    return pathname


if __name__ == "__main__":
    app.run_server(debug=True)
```

Can you test this example at your end? And if possible could you share an example that demonstrates the problem you're seeing?

**Comment by matthijsramlab** [comment]
Hi am having the same problem. For me this examples does not work. I do not why... also downgrading to the previous version does not fix it.

**Comment by matthijsramlab** [comment]
I just downgraded to the Dash version 1.8 that solved my problem. Maybe it has something todo with https://github.com/plotly/dash-core-components/pull/743

**Comment by tcbegley** [comment]
> I just downgraded to the Dash version 1.8 that solved my problem. Maybe it has something todo with plotly/dash-core-components#743

Thanks @matthijsramlab, it does indeed appear this was the problem. I tested against Dash 1.8.0 since that's what the original post found the issue with, but with Dash 1.9.0 I can now reproduce the problem. Will see what's necessary to fix.

**Comment by tcbegley** [comment]
@matthijsramlab @daniel-saunders 

I just made a prerelease that I think will address this issue, if you have time to try it out I'd appreciate knowing if it resolved the issue for you!

```
pip install dash-bootstrap-components==0.8.3rc1
```

I still need to check that this hasn't broken anything else (shouldn't have...) but assuming it's all good I'll try release it as a full release over the next couple days.

**Comment by mkschulze** [comment]
I also had the same issue with latest public release, now it's fixed for me.

dash-bootstrap-components==0.8.3rc1
dash==1.9.0
python==3.8.0

**Comment by beautah** [comment]
confirm as well - 
I also had the same issue with latest public release, now it's fixed for me.

dash-bootstrap-components==0.8.3rc1
dash==1.9.0
python==3.8.0

Thanks!

**Comment by tcbegley** [comment]
Thanks for all the reports. The fix is now part of the latest stable release 0.8.3 (on PyPI,  conda to follow soon).

You should now be able to fix with

```
pip install -U dash-bootstrap-components
```

You can check which version you have installed with

```
python -c "from dash_bootstrap_components import __version__; print(__version__)"
```

I'm going to close this now, but if anyone has further problems feel free to reopen or open a new issue.

**Comment by tstorek** [comment]
Hello,

apparently, the issue is back again. 

Using 

dash                      1.18.1
dash-bootstrap-components 0.11.1
dash-core-components      1.14.1
dash-html-components      1.1.1

I am not able to run @tcbegley example from above. 
Although the url in the browser address bar is updated the dcc.Location callback is not triggered and hence the content not updated automatically. Only after pressing return. Using external_link is a workaround but, in that case the whole page is reloaded instead of only the target component.

However, the example form [here](https://community.plotly.com/t/location-object-update-from-callback-does-not-fire-an-input-location-callback/38355/2) using normal dcc.Link seem to work as expected with only reloading the updated component.

Can we reopen this or is a new issue better?

Using normal dcc.Link seems to work.

**Comment by tcbegley** [comment]
Hey @tstorek 

I just ran the example above without issues using latest versions of dash and dash-bootstrap-components

```
dash                      1.18.1
dash-bootstrap-components 0.11.1
dash-core-components      1.14.1
dash-html-components      1.1.1
```

Do you see any error messages either in the terminal or in your browser's console?

**Comment by tstorek** [comment]
Hello,
no, there is no output at all. 

This is weird. Only the browser address-bar is updated. But not the content. Substituting dbc.NavLink with dcc.Link works as expected. I have no clue where is may come from.

**Comment by tcbegley** [comment]
Here's what I see when I run the app.

![ezgif-7-2278fa948116](https://user-images.githubusercontent.com/15220906/103231012-b17cc700-492e-11eb-99e2-c7e7d61379d5.gif)

You just see an unchanging `/` below?

Which browser / OS are you using?

**Comment by tstorek** [comment]
I just tried FIrefox and Chrome.  I now tested with a fresh project with dash only environment and WOW it works :)

**Comment by tcbegley** [comment]
Huh... that's very odd. If you get any clarity on what might have been going on in the other environment or you see the same thing happening again do let me know.

**Comment by songholmes** [comment]
Hi, actually I want to feedback on the same issue with the latest version of dash and dash-bootstrap for the same script in a conda env. Also an unchanging '/' below.

And I just accidentally tried to install the dash-bootstrap through pip install rather than conda install, the issue gone

So I just suspect, if there is any BUG for the package in the conda env or there is any conflicts for it

**Comment by tcbegley** [comment]
Hey @songholmes 

Can you check for me which versions of Dash and dash-bootstrap-components you had installed in your conda environment? The conda packages are sourced from PyPI by conda-forge so there should be no difference. FWIW I just installed dash-bootstrap-components in a fresh conda environment and tested the above app with no issues.

You can check the versions you have installed either by running

```
conda list
```

or doing something like

```
python -c "from dash import __version__; print(__version__)"
python -c "from dash_bootstrap_components import __version__; print(__version__)"
```

**Comment by songholmes** [comment]
Hi Tom, 
my package version is as below:
  dash: 1.19.0
  dash-bootstrap-components: 0.12.2

actually, I tried again with conda env recently and also not found this issue. 
Anyway, since the issue is not there, thanks for your clarification

**Comment by iwatts3519** [comment]
I am getting this issue now with:
```
dash                                                  1.19.0             pyhd3eb1b0_0
dash-auth                                         1.4.1              pyhd3deb0d_0    conda-forge
dash-bootstrap-components           0.12.2                   pypi_0    pypi
dash-core-components                    1.15.0                   pypi_0    pypi
dash-html-components                    1.1.2                    pypi_0    pypi
dash-renderer                                   1.9.0                    pypi_0    pypi
dash-table                                         4.11.2                   pypi_0    pypi

```
if I change dbc.NavLink to dcc.Link it works perfectly, and I have another app where dash bootstrap components is 0.12.0 and dbc.NavLink works perfectly - I suspect if I downgrade bootstrap components in the current app it will work (That is my next step) but something is wrong here

**Comment by tcbegley** [comment]
Hi @iwatts3519, thanks for reporting

I  am not seeing an issue running the example app above (pasted below again for convenience) with the below versions of Dash libraries installed which appear to match those that you have installed.

Could you both test the simple app and see if it works at your end, and if so, possibly share a version of your app that demonstrates the issue?

Could you also let me know what OS and browser you are using?

```
dash                      1.19.0
dash-bootstrap-components 0.12.2
dash-core-components      1.15.0
dash-html-components      1.1.2
dash-renderer             1.9.0
dash-table                4.11.2
```

```python
import dash
import dash_bootstrap_components as dbc
import dash_core_components as dcc
import dash_html_components as html
from dash.dependencies import Input, Output

app = dash.Dash(external_stylesheets=[dbc.themes.BOOTSTRAP])

app.layout = dbc.Container(
    [
        dcc.Location(id="url"),
        dbc.Nav(
            [
                dbc.NavLink("Page 1", href="/page-1"),
                dbc.NavLink("Page 2", href="/page-2"),
            ]
        ),
        html.Div(id="output"),
    ]
)


@app.callback(Output("output", "children"), [Input("url", "pathname")])
def on_url_change(pathname):
    return pathname


if __name__ == "__main__":
    app.run_server(debug=True)
```

**Comment by iwatts3519** [comment]
Hi
I am happy now to have it running on 0.12.0. When I have some time, possibly tomorrow, I will recreate a copy of the environment, upgrade to 0.12.2 and try my app and your app in the upgraded version. I don't want to mess with it now it is working.
Cheers
Ian

**Comment by iwatts3519** [comment]
Well that is really odd - I made a copy of the folder, rebuilt the Virtual Environment with 0.12.2 (using the environment.yml file from the old environment with just that change from 12.0 to 12.2 in) and it now works with your script and mine - I guess that rebuilding the environment sorted out some dependencies that had maybe got messed about. Anyway, it works :-)

**Comment by iwatts3519** [comment]
Hi
I've come to it again this morning and the problem is back - neither your script nor my script works - nothing has changed in the environment - it isn't a PyCharm issue as I have also run it in Visual Code. I've also tried downgrading to 0.12.0 and the problem still persists. The only thing that works is changing dbc.Navlink to dcc.Link.
I'm using Windows 10 (10.0.19041) and Google Chrome (Version 91.0.4472.77 (Official Build) (64-bit))
I've included my requirements.txt file in case there is another package that is wrong?
[requirements.txt](https://github.com/facultyai/dash-bootstrap-components/files/6591357/requirements.txt)

**Comment by erusto** [comment]
dbc.NavItem(dbc.NavLink("Agregated",active="exact", href="/page-1", external_link=True,)),
include "external_link=True,"

this should work

**Comment by erusto** [comment]
> Well that is really odd - I made a copy of the folder, rebuilt the Virtual Environment with 0.12.2 (using the environment.yml file from the old environment with just that change from 12.0 to 12.2 in) and it now works with your script and mine - I guess that rebuilding the environment sorted out some dependencies that had maybe got messed about. Anyway, it works :-)

dbc.NavItem(dbc.NavLink("Agregated",active="exact", href="/page-1", external_link=True,)),
include "external_link=True,"

this should work

> Hi
> I've come to it again this morning and the problem is back - neither your script nor my script works - nothing has changed in the environment - it isn't a PyCharm issue as I have also run it in Visual Code. I've also tried downgrading to 0.12.0 and the problem still persists. The only thing that works is changing dbc.Navlink to dcc.Link.
> I'm using Windows 10 (10.0.19041) and Google Chrome (Version 91.0.4472.77 (Official Build) (64-bit))
> I've included my requirements.txt file in case there is another package that is wrong?
> [requirements.txt](https://github.com/facultyai/dash-bootstrap-components/files/6591357/requirements.txt)
> 
> EDIT: I should also note that I am using dbc.NavbarSimple, which may or may not be contributing to the problem, but also note that the brand_href attribute likewise doesn't work in returning you to the home page.

dbc.NavItem(dbc.NavLink("Agregated",active="exact", href="/page-1", external_link=True,)),
include "external_link=True,"

this should work

**Comment by Predator83** [comment]
Is there any updates on this issue? I am facing the same problem and using external_link=true actually triggers a complete page refresh; so it is actually a workaround but does not solve the problem. 

I noticed that this issue in my case seems to be caused by an interaction of 2 callbacks:
1st callback like the one in the example above triggering the update of the displayed page, so it has a html.div as Output 
2nd callback is reacting on an URL change but has a doc.storage as output. 

I also noticed that in a different example if you return more than one value by the callback updating the html.div (by return [htmlPage, someOtherValue] the same issue appears.

**Comment by tcbegley** [comment]
Hi @Predator83 

`external_link=True` triggering a page refresh is expected. Navigation in Dash generally works by Dash intercepting the link, updating the address bar, preventing the browser from processing it as a regular link, and then triggering callbacks based on the pathname. Setting `external_link=True` overrides this behaviour and forces the page and hence app to be completely reloaded.

Could you share some code that reproduces the issue you're seeing? I'm happy to try and help debug, but based on your description I don't fully understand what the issue is at the moment.

**Comment by Predator83** [comment]
Hi @tcbegley,

thanks for your quick response. After sleeping a night over the issue I think I found the root cause and it is not connected to the navbar itself directly. 

I have 2 callbacks that are triggered by an URL change: 

1) The callback responsible for updating the navbar
2) The callback responsible for updating a graph on the "home" page 

Essentially what happens is callback 2 tries to update the graph in the "home" page and in parallel callback 1 is triggered and overwrites the "home" page while the graph is still being updated. Result is that it looks like callback 2 is broken when callback 1 is also active.

**Comment by tcbegley** [comment]
Got it, glad you managed to figure it out.
